lib/patchelf/alt_saver.rb

```diff
@@ -0,0 +1,817 @@
+
+    def shift_file(extra_pages, start_page)
+      oldsz = @buffer.size
+      shift = extra_pages * page_size
+      grow_file(oldsz + shift)
+
+      buf_move! shift, 0, oldsz
+      with_buf_at(ehdr.num_bytes) { |buf| buf.write "\x00" * (shift - ehdr.num_bytes) }
+
+      ehdr.e_phoff = ehdr.num_bytes
+      ehdr.e_shoff = ehdr.e_shoff + shift
+
+      @sections.each_with_index do |sec, i|
+        next if i.zero? # dont touch NULL section
+
+        shdr = sec.header
+        shdr.sh_offset += shift
+      end
+
+      @segments.each do |seg|
+        phdr = seg.header
+        phdr.p_offset += shift
+        phdr.p_align = page_size if phdr.p_align != 0 && (phdr.p_vaddr - phdr.p_offset) % phdr.p_align != 0
+      end
+
+      add_segment!(
+        p_type: ELFTools::Constants::PT_LOAD,
+        p_offset: 0,
+        p_vaddr: start_page,
+        p_paddr: start_page,
+        p_filesz: shift,
+        p_memsz: shift,
+        p_flags: ELFTools::Constants::PF_R | ELFTools::Constants::PF_W,
+        p_align: page_size
+      )
+    end
+
+    def sort_phdrs!
+      pt_phdr = ELFTools::Constants::PT_PHDR
+      @segments.sort! do |me, you|
+        next  1 if you.header.p_type == pt_phdr
+        next -1 if me.header.p_type == pt_phdr
+
+        me.header.p_paddr.to_i <=> you.header.p_paddr.to_i
+      end
+    end
+
+    # section headers may contain sh_info and sh_link values that are
+    # references to another section
+    def collect_section_to_section_refs
+      rel_syms = [ELFTools::Constants::SHT_REL, ELFTools::Constants::SHT_RELA]
+      # Translate sh_link, sh_info mappings to section names.
+      @sections.each_with_object({ linkage: {}, info: {} }) do |s, collected|
+        hdr = s.header
+        collected[:linkage][s.name] = @sections[hdr.sh_link].name if hdr.sh_link.nonzero?
+        collected[:info][s.name] = @sections[hdr.sh_info].name if hdr.sh_info.nonzero? && rel_syms.include?(hdr.sh_type)
+      end
+    end
+
+    # @param collected
+    # this must be the value returned by +collect_section_to_section_refs+
+    def restore_section_to_section_refs!(collected)
+      rel_syms = [ELFTools::Constants::SHT_REL, ELFTools::Constants::SHT_RELA]
+      linkage, info = collected.values_at(:linkage, :info)
+      @sections.each do |sec|
+        hdr = sec.header
+        hdr.sh_link = find_section_idx(linkage[sec.name]) if hdr.sh_link.nonzero?
+        hdr.sh_info = find_section_idx(info[sec.name]) if hdr.sh_info.nonzero? && rel_syms.include?(hdr.sh_type)
+      end
+    end
+
+    def sort_shdrs!
+      section_dep_values = collect_section_to_section_refs
+      shstrtab_name = @sections[ehdr.e_shstrndx].name
+      @sections.sort! { |me, you| me.header.sh_offset.to_i <=> you.header.sh_offset.to_i }
+      update_section_idx!
+      restore_section_to_section_refs!(section_dep_values)
+      ehdr.e_shstrndx = find_section_idx shstrtab_name
+    end
+
+    # given a +dyn.d_tag+, returns the section name it must be synced to.
```

Comment:
Nit: Update this to note that it can return `nil`, per the comments in the body.